Improve Pixelblaze, results summary, various

JohnnyMarnell opened the pull request:
Hello! Thanks so much for creating this tool! 🫶🏻

The main changes are streamlining the Pixelblaze backend process, and offering a summary CLI (the logs and CV processes can be tough to follow).

Changes:
* Streamlining Pixelblaze: no longer require uploading pattern steps, use (cached) compile and switch APIs
* Implement `set_leds()` for Pixelblaze backend for inter-scan feedback
* Auto discover Pixelblaze IP and connect (or use existing `--server` option)
* Add a results / summary command that looks at 2D scans and 3D merge and summarizes what is missing
* Add some documentation for more details with steps, especially Pixelblaze, camera settings and process
* Add way to set and persist window positions and sizes

JohnnyMarnell committed:
simplify pixelblaze steps, automatic pattern render

JohnnyMarnell committed:
logs are pretty tough to follow, add simple script to assess what's been done

JohnnyMarnell committed:
vibe code simpler DLT + SVD solver, erring on completeness

- COLMAP + RANSAC are amazeballs, but they require enough complacent input data.
- This is a lighter weight similar solver using Direct Linear Transform and triangulation by Single Value Decomposition that will fill in any missing LEDs that have at least 1 2d scan with at least 6 common points with existing 3d
- For LEDs (as opposed to complicated 3d images), this should work pretty well in practice (says: Robots)

JohnnyMarnell committed:
wrap in project cli style

JohnnyMarnell committed:
simplify text displays

JohnnyMarnell committed:
minor log cosmetics

JohnnyMarnell committed:
first pass at estimated camera logging, looks wrong

JohnnyMarnell committed:
improve the relative camera estimated logs, use azimuth + elevaishe

JohnnyMarnell committed:
Burn more forests using more AI for improved math

JohnnyMarnell committed:
Small doc addition

JohnnyMarnell committed:
Simplify pattern loading even more, impl set_leds()

JohnnyMarnell committed:
big improvements: pixelblaze auto discover, fully fix set_leds() + mapping iterations

JohnnyMarnell committed:
persist window size and location, yeesh cv2 / this window system sucks for this

JohnnyMarnell committed:
windows...

JohnnyMarnell committed:
file based window state

JohnnyMarnell committed:
Improve docs

JohnnyMarnell committed:
Merge branch 'TheMariday:main' into pb-revisit-2

JohnnyMarnell committed:
Merge branch 'pb-revisit-2' of https://github.com/JohnnyMarnell/marimapper into pb-revisit-2

JohnnyMarnell committed:
Undo window size + position attempt nightmare

JohnnyMarnell committed:
Introduce way to set window positions and sizes

JohnnyMarnell committed:
run linting

TheMariday commented:
Hi @JohnnyMarnell, you are more than welcome and thank you for the PR! Woah this is a big-un!

It'll take me a while to review this and also get my hands on a pixelblaze to test your changes here.

The only thing that's jumping out at me is the addition of a second separate photogrammetric solver.

I'm curious as to how colmap failed your use case!

JohnnyMarnell committed:
rm alternate solver, fix 3.9 errors

JohnnyMarnell commented:
Hey hey, thanks so much for taking a look (and so fast).

Kk, I removed the extra solver code that I (well, AI mostly...) had. I did like that it could run fast and self-contained, and had more tolerances to play with, but, to be fair, I didn't verify its results heavily, and, more importantly, hadn't ever tried the two args I see in main marimapper command: `--interpolation_max_fill` and `--interpolation_max_error`. Unsurprisingly, these do get me past the problem I've consistently had: doing many scans that do come up with all LEDs detected across them, but 1 to a few missing from the 3D output. Not sure how its interpolation methods would compare to the multiple best effort and fallback methods that were in that code.

And hopefully all CICD errors are fixed now (I'm kind of guessing at what commands to run locally...).

TheMariday commented:
The interpolation argument is a post processing step that estimates missing LEDs based on their ID's in 3D space so is after the solver has resolved it's state. I'm glad you found it as an option but your implementation of a secondary, more lenient solver does bring something to mind:

COLMAP does have a few thresholds for maximum error including a threshold for the RANSAC optimiser.

I'm wondering if it would be good to add an argument which is basically like `--best_effort` which sets those thresholds to like a million and leave it to the user how much they want to threshold the output after the solver has completed.

I also like the idea of a post processing CLI to do things like rotations and statistics as I agree that the logs are hard to follow. Hard to get multiprocessing working nicely with a CLI on the best of days!

I also really appreciate the changes to the pixelblaze code, I think I want to test these thoroughly as if they work consistently across platforms then it would be a real treat to have auto-discovery and colour compatibility!

Thank you again for this and I am blown away with the amount of work and the entire second photogrammetry engine! Absolute madness. Also I would love to know if there are any other pain-points you had that I could take a look at. The window position is a fab addition!

I also hope it's ok if I make some changes to this branch before putting it in? I want to add some unit tests, bump the version number and tidy up the docs a bit.

If you'd like to take this privately, you can of course drop me a line at t.me/themariday.

I'll also add some more information on a developer guide so to speak for things like the style checks. For now you can see the workflows in the .github folder and mimic those. 

This should do for now though:
```
pip install flake8 flake8-bugbear black
flake8 .
black .
```

Commit:
Literally one f string out of place, that's all!

TheMariday commented on marimapper/detector.py:
```diff
@@ -89,7 +90,24 @@ def draw_led_detections(image: cv2.Mat, led_detection: Optional[Point2D]) -> np.
+
+    target_width = int(target_height * aspect_ratio)
+
+    if not getattr(show_image, "setup_done", False):
```
I had no idea you could assign attributes to functions.

I am not sure how I feel about this :D

https://sethdandridge.com/blog/assigning-attributes-to-python-functions

TheMariday replied:
I think this could also be achieved by checking `cv2.getWindowProperty(window_name, 0)`

Also does this just set it to a fixed size then write that to disk? I can't find if it updates the cached values with new values if the user moves / rescales it.

JohnnyMarnell replied:
Ya, you can do pretty much anything with python, for better and worse... I was rushing, I think a static var lazy init kind of marker would be better? Feel free to do whatever style you'd prefer, but it would be great if the functionality / end result stayed roughly the same. I spent a lot of frustrating time trying to get this to work, in my experience, cv2 and o3d have been difficult. Per your other comments, I also would've preferred that it would remember resizing / arrangement and automatically and use that. That's what I started with and sunk a lot of time into trying to get it to work. Going back to multiprocessing / overall architecture, this is an up hill battle. I think I read also that you'd need to upgrade o3d version, the current one doesn't offer any feedback into user movements (unverified). I *kind of* had cv2 window working, but the scaling factors / pixel sizes weren't even lining up between the two windows. And I imagine it's going to be even more headache getting it to work for multiple platforms.

Also the instantiation and order of events seems inconsistent and non-deterministic (hence the lazy init / `setup_done` approach). At the end of the day, it's frustrating to have to move and resize a giant window every run, so I at least hope this method (optionally defining a json file) can live on?

TheMariday commented on marimapper/scripts/results_cli.py:
```diff
@@ -0,0 +1,101 @@
+import click
```
If we can modify this script to not use `click` or `pandas` that would be amazing as the install is already quite slow from the current dependencies.

This could also be sped up with proper packaging and upload to pypi, but that's a lotta work!

JohnnyMarnell replied:
Not using these won't speed up install, as click and pandas are already transitively required by other libraries (check `pip tree`). They're good, common libraries and already transitively imported, so I actually think it's worth using them (and pandas, in addition to CSV helpers, has great helpers for textual display, like easy to read tables).

TheMariday commented:
Also I'm adding some comments here, please don't feel pressured to address them as you've done loads. They're mostly notes for myself.

TheMariday commented on CLAUDE.md:
I'd quite like to remove this file, as I am quite staunchly anti-AI and don't really want to encourage people to use it on this repo.

I've had some, rather questionable logic in some other AI-powered contributions...

JohnnyMarnell replied:
Personally, I think being staunchly-pro or anti is a mistake, it's good and bad at things, and here whether we like or not. I definitely agree it can create questionable stuff if we're not paying attention. Removing though!

TheMariday commented on .gitignore:
```diff
@@ -160,4 +160,10 @@ cython_debug/
\ No newline at end of file
+.idea/
+
+# uv is referenced, but not primary package management system
```
UV is the primary package / venv management system for Marimapper?

JohnnyMarnell replied:
If that's the case, then uv.lock should already be in source control, I was surprised it wasn't. I'll remove this and check it in.

TheMariday commented on README.md:
```diff
@@ -154,6 +158,8 @@ If you want a lot more reading material, run `marimapper` with `-v` to put it in
 This will tell you pretty much everything marimapper is doing under the hood.
 Also good if you're just curious as to why *x* is taking so long!
 
+See also [here](docs/backends/PixelBlaze.md) for some more example walk through of the process.
```
Perhaps we should break this out into a new md which is something like tips and tricks or troubleshooting to keep the pixelblaze specific stuff separate from the general exposure advice.

JohnnyMarnell replied:
That sounds fine to me, is that something you'd want to do or should I? You have a lot more experience using this great tool, and a vision I'm assuming.

TheMariday commented on docs/backends/custom.md:
```diff
@@ -23,6 +23,12 @@ class Backend:
         # else:
         #     some_led_library.set_led(led_index, (0, 0, 0))
+    
+    def set_leds(self, buffer: list[list[int]]) -> None:
```
Now this is available for pixelblaze, I feel like this should be more of the "primary" way that LEDs are controlled with a backup "simple" method and there should be some more tests and demo examples of how the colouration works.

JohnnyMarnell replied:
I both think that makes sense and things could be simpler, but also think doing anything comes with risk. Up to you

TheMariday commented on marimapper/backends/pixelblaze/pixelblaze_backend.py:
```diff
@@ -10,22 +14,81 @@
+                # Try beacon scan on network
+                devices = discover_pixelblazes(timeout=3.0)
+                if devices:
+                    pixelblaze_ip = devices[0]
```
We might need to cover what we do if we find multiple devices here.

JohnnyMarnell replied:
Really don't think this will ever be a problem in practice, and there's always `--server`, and I looked again and see I'm already logging how many were discovered?

TheMariday commented on marimapper/utils.py:
```diff
@@ -49,3 +52,37 @@ def __exit__(self, *_):
+
+
+CONFIG_DIR = Path.home() / ".config" / "marimapper" / "windows"
+_CACHE = {}
```
If `position_window` is only hit twice and we check first if the window has been initialised, it might be best to remove this cache due to \*lighting cracks\* *cache invalidation!*

JohnnyMarnell replied:
See previous comments, working with CV2 and O3D windows, like the various multiprocessing stuff, is kind of a nightmare with current state. I think imshow() is called a ton, so I'm not married to how it works, just want to be able to clunkily set some kind of window position state, and have it not read the fs every time.

JohnnyMarnell commented:
> The interpolation argument is a post processing step that estimates missing LEDs based on their ID's in 3D space

FWIW, running again with different arguments changed every position line in the CSV, which was not what I would expect. (I didn't look deeply into by how much, I did quickly visually compare and they looked pretty similar).

> I'm wondering if it would be good to add an argument which is basically like --best_effort

I think in general it would be nice to expose more config for COLMAP.

> I also like the idea of a post processing CLI to do things like rotations and statistics as I agree that the logs are hard to follow. Hard to get multiprocessing working nicely with a CLI on the best of days!

Yea, this would be nice, and part of what I was going for. The camera/frustum drawing in the open 3d viewer is cool, but something textual / summary would be nice too, to verify camera movements and adjustments are seeming to line up. I'm glad you mentioned multiprocessing, I've found this working with this part of the code very tough :( (I think it's especially challenging on Mac, I haven't run any other platforms, but I run into a lot of hangs with no logs, camera stops working, etc). I think it could use some refactor/redesign with how these processes work and talk to each other. But ofc that would take significant effort. 

> I also hope it's ok if I make some changes to this branch before putting it in? I want to add some unit tests, bump the version number and tidy up the docs a bit.

Yes please these would all be great!

FWIW, there can be python 9 errors (like pipe for union vs `Union`) that flake8 and black commands didn't catch.

JohnnyMarnell committed:
Address CR feedback

JohnnyMarnell committed:
Merge branch 'pb-revisit-2' of https://github.com/JohnnyMarnell/marimapper into pb-revisit-2

TheMariday commented:
Heyo! So I've been having a read and a think.

# COLMAP settings

Absolutely lets get those exposed for advanced users. I need to do some "best effort" tests. There are however about half a billion COLMAP options so I'll have to do some testing to see what gives the effect we want.

# Non-determinism of led interpolation

I've had a look at the source code and I'm not aware of how the `fill_gaps` method could me adjusting the other LED positions. I am wondering that this might be to do with the general non-determinism of the state optimiser, however I've found it relatively stable in my test.

# Click and Pandas
Firstly, I can see that Click is being included via O3D anyway so yes that would be good to use! I'm wondering if we could then replace the current argparser with Click for the rest of the CLI as I'm reticent to have two different cli frameworks in the same tool.

However I can't see where Pandas is being pulled in. It would be nice to have for the CSV parsing and data table visualisation, however it feels a bit bloaty for just those two use cases on top of drawing in 3 other dependencies.

*EDIT ^ This is wrong, for some reason I had a problem my end and pandas is being pulled in.*


# Window resizing.
So if we [upgrade to the new Open3D gui](https://github.com/TheMariday/marimapper/issues/59), we should be able to fetch both the OpenCV and Open3D window sizes via an on-close callback which can then write it to disc ready to be loaded on initialisation. OpenCV does though pose a challenge, when a user re-scales a window, it doesn't respect the original aspect ratio. There doesn't seem to be a way around this, or at least none that I could find. I think your solution here of just having some more "sensible" defaults might be the way forward.

# General multiprocessing complexity

You raised some really good points about the multiprocessing code being confusing to get your head around. I have hit many a weird hang issue whilst developing it and it's lack of hard synchronisation can make initialisation hard to debug.

I'm not sure if you were suggesting changes to readability and docs or a complete overhaul but I would love to hear how you might approach this differently.

The main constraint that I'm under is that I want to be able to see the reconstruction happening live, Open3D //must// run in the main thread and so does OpenCV.

# Statistics

I'm also curious if you've got any ideas as to how to report logs and statistics better, I like the post-processing tool but would love to see it liv. One idea for this is we append statistics to our reconstructions and have 2 types, "live" reconstructions, which just constantly push updated LED positions back to the visualiser and "final" reconstructions which are triggered by the capture thread once a scan has been completed. The CLI could then stall until that final reconstruction has been completed and then print out the statistics.
This would however slow down large reconstructions as each scan would have to wait for one extra reconstruction per camera view. It might be worth it however. I'm doing my best to make sure this tool works for tens of leds up to tens of thousands. (Well my use case is just over nine thousand.)

# Final notes

Also thank you so much for your input on this so far, there's absolutely no pressure to respond if you've got better things to do. I'm so honoured that people want to work on this with me at all to be honest as yea it's a bit of a mess! :D

TheMariday commented:
Actually, hell we might be able to ditch the multi-window thing and move straight to the new Open3D visualiser GUI. I'll do some research

TheMariday commented:
I've just had a brainwave. Don't worry about anything I've mentioned above, I have an idea, I'll get back to you.

TheMariday commented:
Hi Jonny.

Thank you for your patience and my apologies for being all-over-the-board.

Your comments about multiprocessing got me thinking, and I think it's time to ditch the whole GUI / CLI in favour for something better. I've found a rather promising library called Viser which can handle all the visualisations, gui, import / export etc. This would also mean that the entire tool could run on two threads. No more OpenCV windows, Open 3d, queue nonsense, hidden cli features, weird hangs, etc etc.

As I'm imminently going to be stripping out all the GUI elements, are you  comfortable stripping this PR down to the changes to PixelBlaze and the documentation? I'd prefer to keep the change set small so we can merge it in sooner rather than later.

I'm currently in the process of getting my hands on a PixelBlaze board so I can test your changes or find someone who can. As soon as I have, I'll get this tested and merged in.

Thank you for inspiring me to improve this tool

Mariday

JohnnyMarnell committed:
set brightness on pixelblaze, fix summary bug

JohnnyMarnell committed:
remove click and pandas use

JohnnyMarnell committed:
remove my dumbness and rushing for window nightmares, vastly simplified

JohnnyMarnell commented:
howdy howdy. I took a look at viser, it does look like a nice, promising library.

as for stripping down the PR, can you take one more look at my latest? overhauling the GUI and run system I imagine will take some time, and I think I've made valuable improvements that it'd be a shame to dismiss in the meantime.

To help address feedback / move my case forward, I've removed pandas and click usage, as well as cleaned up that results summary. (I still think this is very useful, and would prefer not to have to maintain a fork / keep that option from others). Here's sample output for my 500 LED Christmas tree attempts. Without this code, I have no idea how successful my scans are, without painfully trying to dig through them:
https://gist.github.com/JohnnyMarnell/8075296684b0d41b5eb5a95f2c1f9f7e

And I was being kind of dumb before with the window stuff. I've cleaned that up a lot, I hope you find it less irking and that it reads much clearer. All it really is meant to be, is a way to manually store window position and sizes via optionally creating a settings file once. Then I don't have to continually move and resize an enormous window running off the screen every time I run. (In the process, it's a lot more responsive for me to mouse movements btw, I believe the cv2.resize() call actually helps).

JohnnyMarnell committed:
run lint

JohnnyMarnell committed:
reeeeeeally need clear, simple, deterministic lint check / command here :(

TheMariday commented:
Hi @JohnnyMarnell, just to keep you in the loop. I'm currently waiting to aquire some Pixelblaze boards to test your changes.

Looking forward to getting that merged along with the window position fixes and documentation.
